File: code/java/easy/0733-floodFill.java
// LGSI interview question

class Solution {
    class Pair {
        int row;
        int col;

        Pair(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    public boolean isValid(int[][] image, int row, int col, int color) {
        return row >= 0 && row < image.length && col >= 0 && col < image[0].length && image[row][col] == color;
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        //add pixel to queue

        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(sr, sc));
        int[][] directions = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };

        //add adjacent pixel if they're same
        while (!q.isEmpty()) {
            Pair curr = q.remove();

            if (image[curr.row][curr.col] == color) {
                continue;
            }
            int temp = image[curr.row][curr.col];
            image[curr.row][curr.col] = color;
            //case: not first time we encounter this pixel
            

            for (int[] dir : directions) {
                int dx = curr.row + dir[0];
                int dy = curr.col + dir[1];

                //add if its valid
                if (isValid(image, dx, dy, temp)) {
                    q.add(new Pair(dx, dy));
                }
            }

        }

        return image;
    }
}

File: code/cpp/easy/0733-floodFill.cpp
class Solution {
    public:
     bool is_valid(vector<vector<int>>& image, int& row, int& col, int& color) {
         return row >= 0 && row < image.size() && col >= 0 &&
                col < image[0].size() && image[row][col] == color;
     }
 
     vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc,
                                   int color) {
         // add the pixels to the queue
         // add it's adjacent to the queue
         // remove from queue and color it
         // case 1: 1st time we come
         // case 2: not 1st time
         queue<pair<int, int>> q;
         q.push({sr, sc});
         pair<int, int> directions[] = {{1, 0}, {-1, 0}, {0, -1}, {0, 1}};
 
         while (!q.empty()) {
             auto curr = q.front();
             q.pop();
 
             if (image[curr.first][curr.second] == color) {
                 continue;
             }
 
             int temp_color = image[curr.first][curr.second];
             image[curr.first][curr.second] = color;
 
             for (auto dir : directions) {
                 int dx = curr.first + dir.first;
                 int dy = curr.second + dir.second;
 
                 if (is_valid(image, dx, dy, temp_color)) {
                     q.push({dx, dy});
                 }
             }
         }
 
         return image;
     }
 };